fuse/daemon.cpp

```diff
@@ -207,14 +217,23 @@ static Special is_special(const char *path) {
 	}
 }
 
-static int exit_attempts = 0;
+// If exit_attempts is > 0, we are in the impossible-to-stop process of exiting.
+// On a clean shutdown, exit_attempts will only ever be increased if context.should_exit() is true.
+static volatile int exit_attempts = 0;
 
-static void cancel_exit()
+// You must make context.should_exit() false BEFORE calling cancel_exit.
+// Return of 'true' guarantees the process will not exit
+static bool cancel_exit()
```

Comment:
I'm getting kind of confused by these comments, possibly due to me not knowing the bigger picture of how this code is run. If the process of exiting is "impossible to stop," then what is the purpose of a `cancel_exit()` function?

Is there a brief period of time where you may be "thinking of exiting," initiating the beginning of an exit, but not get far enough to cause the `exit_attempts` counter to increment, which indicates that it is still safe to abort the exit? In other words, you can start to exit, but you can always back out of it until you've hit the point of no return, which is indicated by incrementing the the `exit_attempts` variable?

This interpretation is the only one that I can come up with that makes your code make sense, but in general, I think you have some specific semantics in your head that aren't reflected in the names of the functions/variable and/or don't have any documentation. Perhaps you could slightly alter the wording of the comment you added to elaborate on this, and that clearly indicate that exits can still be canceled up until the point that the `exit_attempts` variable has been incremented?

Reply:
Yes, there is a not-so-brief period (2s) where the server is not in-use, but exit has not started. The server stays alive a little bit assuming you might start another job soon. When all uses are done, we call `schedule_exit()` to trigger exit in 2 seconds. If another use appears, we `cancel_exit()`, but have to be careful to distinguish whether or not we've already begun to exit.

Reply:
The server can be three states: in-use, idle, exit-started. There is a 2s delay between idle and exit started. Actually exiting also takes time, depending on how busy your system is, maybe also on the order of seconds (normally a few ms, tho).

Reply:
Rereading the comments and function names, it seems to me the code correctly reflects this... and you also deduce the correct big picture from those comments, without reading how the rest of it works, so I'd say all is well.